Allow secondary user backup to USB

Uldiniad opened the pull request:
By default, Android exposes USB devices only to the main user.
In order to query, read and write to it, the signature permission INTERACT_ACROSS_USERS_FULL (optional) is granted to create Seedvault's context as the system user.

Change-Id: I0b1b4c8c5aeeb226419ff94e15f631ebe1db66df
Issue: calyxos#437
Issue: https://github.com/seedvault-app/seedvault/issues/77

Uldiniad committed:
Allow secondary user backup to USB

By default, Android exposes USB devices only to the main user.
In order to query, read and write to it, the signature permission INTERACT_ACROSS_USERS_FULL (optional) is granted to create Seedvault's context as the system user.

Issue: calyxos#437
Issue: https://github.com/seedvault-app/seedvault/issues/77
Change-Id: I0b1b4c8c5aeeb226419ff94e15f631ebe1db66df

grote commented on app/src/main/java/com/stevesoltys/seedvault/ui/storage/RestoreStorageViewModel.kt:
```diff
@@ -31,8 +32,6 @@ internal class RestoreStorageViewModel(
             }
             if (hasBackup) {
-                saveStorage(uri)
-
```
You are saving the storage off the I/O thread here, possibly on the UI thread.

However, I don't understand this change in general. If there's an issue here, we should maybe look into it separately. Or is it only your changes in this PR that are surfacing an issue that otherwise doesn't exist?

Uldiniad replied:
There were errors cause by the code before it trying to read the value and it not being present in `SettingsManager.getStorage`. I don't have the log currently. Additionally the BackupStorageViewMode (https://github.com/seedvault-app/seedvault/blob/android12/app/src/main/java/com/stevesoltys/seedvault/ui/storage/BackupStorageViewModel.kt#L35)l stores it early as well.
It is very likely tied to this PR and not a general issue

grote replied:
First of all, saving the storage off the I/O thread here, possibly on the UI thread is an error that should get resolved in any case. No disk I/O on the UI thread! (If this is happening in other places of the code, that's an error there as well and should get fixed eventually.)

Then, we can't simply move this call around without fully understanding the implications. Just because the backup flow does it, doesn't mean we can do it here as well. If the logic is the same in both flows, why aren't both view models sharing this in common code in `StorageViewModel` then? Just a programming oversight?

I myself don't remember, but looking just at this method here I wonder, what if there are no backups in the location the user has chosen? Will it always be possible to try a different location even if the first one was already saved? Did you try that?

Maybe we don't even need to move the saving call. If we knew what the issue was here, we might be able to come up with a different solution.

grote replied:
It looks the only issue that you might have run into is in the initialization of `DocumentsProviderStoragePlugin` which tries to access `storage` at construction time now (which might actually cause issues in other places if storage is not yet available). However, if `storage` is still `null` there, then there shouldn't be any crashes or anything. But it might have happened that you couldn't access the system user USB drive from a work profile at restore time.

Uldiniad replied:
Moved it back. Will post log when I have it

chirayudesai replied:
Bump.

@Uldiniad do one thing, split this part into a separate commit - just this one liner.

Uldiniad replied:
Done

grote commented on app/src/main/java/com/stevesoltys/seedvault/ui/storage/StorageRootResolver.kt:
```diff
@@ -34,6 +35,15 @@ internal object StorageRootResolver {
                     if (root != null) roots.add(root)
                 }
             }
+            if (UserHandle.myUserId() != UserHandle.USER_SYSTEM) {
```
Do we need to do this for *all* authorities or could we also check the current authority and only query with system context for authorities that USB storage is using?

Uldiniad replied:
I am ok with limiting this to USB authorities only, but I don't know how we can identify them

grote replied:
Have you tried checking the existing log or adding explicit authority logging here looking for a pattern to the authority? I'd imagine it to always be the same one e.g. `com.android.externalstorage.documents`.

grote replied:
If we do this only for a specific authority, maybe this isn't even the best location to do it as this code is called for all authorities. Maybe an earlier place in the calling hierarchy of this method would be more appropriate?

grote replied:
Could factor this `if` part out into its own method and call this in the loop in `getStorageOptions()` only if the authority matches the one we care about.

Uldiniad replied:
Would this negatively affect `openDocumentTree` in `StorageActivity` since it also calls `getStorageRoots()` ?

grote replied:
Good point, that's the fall-back for selecting storage without MANAGE_DOCUMENTS permissions. It only works on specific Uris selected by the user though. So how would the work-profile user select system usb storage via the file picker? If that's not possible anyway, I think we can ignore this issue.

Uldiniad replied:
I just tried attaching a USB and adding an attachment from the USB using the file picker to a conversation on the messaging app in the work profile and it worked. Seems like this usecase needs handling

grote replied:
Wow, so the files app has code for this as well? So you can access USB devices from work profile after all?

Uldiniad replied:
Yeah. There are various intents that work across profiles. USB_DEVICE_ATTACHED is one of them. Additionally the files app has some internal code that limits exposure to other profiles.
So we have to handle this within the getStorageRoots function given this scenario?

Uldiniad replied:
Added check for `com.android.externalstorage.documents`

grote commented on app/src/main/java/com/stevesoltys/seedvault/ui/storage/StorageRootResolver.kt:
```diff
@@ -34,6 +35,15 @@ internal object StorageRootResolver {
+                val c: Context = context.createContextAsUser(UserHandle.SYSTEM, 0)
+                c.contentResolver.query(rootsUri, null, null, null, null)?.use { cursor ->
+                    while (cursor.moveToNext()) {
+                        val root = getStorageRoot(context, authority, cursor)
```
`getStorageRoot` root is called with the current user context, not the system context. Is that intentional? Could that have worked in your testing?

Uldiniad replied:
This context is sent to query package manager for icons. You mentioned you'd prefer using the context of the current user for package manager?

grote replied:
Ah ok, it wasn't obvious from the diff that the context is only used to fetch the icon. That's fetching a drawable from the current app in this case, so it probably should use app context, but I'd add a comment explaining why using the app context here and not `c`.

Uldiniad replied:
Done

grote commented on storage/lib/src/main/java/org/calyxos/backup/storage/plugin/saf/SafStoragePlugin.kt:
```diff
@@ -30,11 +30,12 @@ private const val TAG = "SafStoragePlugin"
 
     private val cache = SafCache()
     protected abstract val root: DocumentFile?
+    protected abstract val context: Context
```
Please add JavaDocs to both contexts. This is a library and it isn't obvious what those contexts are and where there are two of them.

Uldiniad replied:
Done

grote reviewed (changes requested):
Looks good. I found some minor stuff and will do one last pass before merging, checking all the context usage in the places you touched to make sure we are not accidentally using the wrong context somewhere.

Uldiniad commented:
```
02-16 16:35:59.761 12887 12990 E DocumentsStorage: Error finding file blocking
02-16 16:35:59.761 12887 12990 E DocumentsStorage: java.io.IOException
02-16 16:35:59.761 12887 12990 E DocumentsStorage:      at com.stevesoltys.seedvault.plugins.saf.DocumentsStorageKt$getLoadedCursor$2.invokeSuspend(DocumentsStorage.kt:281)
02-16 16:35:59.761 12887 12990 E DocumentsStorage:      at com.stevesoltys.seedvault.plugins.saf.DocumentsStorageKt$getLoadedCursor$2.invoke(Unknown Source:8)
02-16 16:35:59.761 12887 12990 E DocumentsStorage:      at com.stevesoltys.seedvault.plugins.saf.DocumentsStorageKt$getLoadedCursor$2.invoke(Unknown Source:4)
02-16 16:35:59.761 12887 12990 E DocumentsStorage:      at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturnIgnoreTimeout(Undispatched.kt:102)
02-16 16:35:59.761 12887 12990 E DocumentsStorage:      at kotlinx.coroutines.TimeoutKt.setupTimeout(Timeout.kt:78)
02-16 16:35:59.761 12887 12990 E DocumentsStorage:      at kotlinx.coroutines.TimeoutKt.access$setupTimeout(Timeout.kt:1)
02-16 16:35:59.761 12887 12990 E DocumentsStorage:      at kotlinx.coroutines.TimeoutKt.withTimeout(Timeout.kt:31)
02-16 16:35:59.761 12887 12990 E DocumentsStorage:      at com.stevesoltys.seedvault.plugins.saf.DocumentsStorageKt.getLoadedCursor(DocumentsStorage.kt:279)
02-16 16:35:59.761 12887 12990 E DocumentsStorage:      at com.stevesoltys.seedvault.plugins.saf.DocumentsStorageKt.getLoadedCursor$default(DocumentsStorage.kt:278)
02-16 16:35:59.761 12887 12990 E DocumentsStorage:      at com.stevesoltys.seedvault.plugins.saf.DocumentsStorageKt.listFilesBlocking(DocumentsStorage.kt:204)
02-16 16:35:59.761 12887 12990 E DocumentsStorage:      at com.stevesoltys.seedvault.plugins.saf.DocumentsStorageKt.findFileBlocking(DocumentsStorage.kt:248)
02-16 16:35:59.761 12887 12990 E DocumentsStorage:      at com.stevesoltys.seedvault.plugins.saf.DocumentsProviderStoragePlugin.hasBackup(DocumentsProviderStoragePlugin.kt:82)
02-16 16:35:59.761 12887 12990 E DocumentsStorage:      at com.stevesoltys.seedvault.ui.storage.RestoreStorageViewModel$onLocationSet$1.invokeSuspend(RestoreStorageViewModel.kt:28)
02-16 16:35:59.761 12887 12990 E DocumentsStorage:      at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
02-16 16:35:59.761 12887 12990 E DocumentsStorage:      at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:241)
02-16 16:35:59.761 12887 12990 E DocumentsStorage:      at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:594)
02-16 16:35:59.761 12887 12990 E DocumentsStorage:      at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60)
02-16 16:35:59.761 12887 12990 E DocumentsStorage:      at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:740)
```
this is the error when save storage is not executed

Uldiniad committed:
Save storage unconditionally on location set

The storage plugin cannot find USB backups for secondary users if it is not saved before checking if backups exist

grote committed:
Save restore storage only when it had backups

Incidentally this gets rid of the need to pass an implementation-specific Uri to the StoragePlugin.

grote committed:
Clean up system USB storage feature a bit

grote commented on app/src/main/java/com/stevesoltys/seedvault/ui/storage/StorageRootResolver.kt:
```diff
@@ -34,6 +36,21 @@ internal object StorageRootResolver {
+                authority == AUTHORITY_STORAGE && UserHandle.myUserId() != UserHandle.USER_SYSTEM
+            }
+            // only proceed if we really got a different [Context], e.g. had permission for it
+            if (context !== c) {
```
I am not a 100% sure that this will work as expected. Mainly did it like this, so we don't run into a situation where the permission isn't granted and it uses the same context creating duplicate roots.

chirayudesai replied:
Seems to work fine.
